src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private T[] grow(int oldSize) {
+        int newSize = oldSize + oldSize / 2;
+        T[] newElementData = (T[]) new Object[newSize];
+        System.arraycopy(elementData, 0, newElementData, 0, elementData.length);
+        return newElementData;
+    }
+
+    public void checkNonExistentOrNegativePosition(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element on this position");
+        }
+    }
+
+    public void checkIndexForAddElement(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element on this position");
+        }
+    }
+
+    public void addElementInsideArrayByIndex(T value, int index) {
+        T[] newElementData = (T[]) new Object[size + 1];
+        if ((size + 1) > elementData.length) {
+            grow(size + 1);
+        }
+        if (index == 0) {
+            System.arraycopy(elementData, 0, newElementData, 1, size);
+            elementData = newElementData;
+            elementData[index] = value;
+            size++;
+        } else {
+            System.arraycopy(elementData, 0, newElementData, 0, index);
+            System.arraycopy(elementData, index, newElementData, index + 1,
+                    size - index);
+            elementData = newElementData;
+            elementData[index] = value;
+            size++;
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAddElement(index);
+        if (index == elementData.length) {
+            elementData = grow(size);
+        }
```

Comment:
you don't need special check for add, you can use the usual one, you just need to place it after the if, it will filter out the case that makes them different (index == size)

Reply:
Unfortunately it doesn't work... This logic works only with special check (index > size || index < 0)...

Reply:
<img width="944" alt="checkIndexForAddElement" src="https://user-images.githubusercontent.com/123567653/230607619-925269cd-93a8-42ab-953c-beba66f44816.PNG">